feat(aws-arch): initial commit

JeremyJonas opened the pull request:
Creates new package `@aws-prototyping-sdk/aws-arch` (**EXPERIMENTAL**)

---
Library to provide metadata for AWS Services and AWS Resources.

This package generates mappings between [@aws-cdk/cfnspec](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/cfnspec) and [AWS Architecture Icons](https://aws.amazon.com/architecture/icons/), and resolves resource metadata between these systems to infer a normalized definition of an `AwsService` and `AwsResource`.

The primary aim of this package is to provide a consistent mechanism for other libraries to retrieve naming and assets associated with CloudFormation resources, and by extension CDK resources.

JeremyJonas committed:
feat(aws-arch): initial commit

agdimech commented on packages/aws-arch/README.md:
````diff
@@ -0,0 +1,90 @@
+```
+
+```ts
+import { AwsArchitecture } from "@aws-prototyping-skd/aws-arch";
````
@aws-prototyping-sdk*

agdimech commented on packages/aws-arch/README.md:
````diff
@@ -0,0 +1,90 @@
+```ts
+import { AwsArchitecture } from "@aws-prototyping-skd/aws-arch";
+
+const s3Bucket = AwsArchitecture.getResource("AWS::S3::Bucket");
````
It would be awesome if these resource types where represented as an enum :) Do you know if the CDK already has this in one spot?

JeremyJonas replied:
The input for `getResource` is typed for all CfnResourceTypes parsed from CfnSpec.

<img width="797" alt="image" src="https://user-images.githubusercontent.com/464119/199877973-5dae4959-f07e-463d-83b6-f49ae7fbe241.png">

Additionally, this package exposes the generated internals as well, which as this as a Dictionary of Service=>Resource=>CfnResourceType
<img width="722" alt="image" src="https://user-images.githubusercontent.com/464119/199878143-7f9bca9f-0c94-414f-8f9b-f87b5f07375a.png">

```ts
import { aws_arch } from '@aws-prototyping-sdk/aws-arch'

AwsArchitecture.getResource(aws_arch.CfnSpec.ServiceResourceDictionary.S3.Bucket)
```

> Actually just noticed the static typing was enforced for this api, but was for the AwsResource.getResource api, I will fix this

JeremyJonas replied:
Discussed offline - resolving

agdimech commented on packages/aws-arch/README.md:
````diff
@@ -0,0 +1,90 @@
+```ts
+const s3Bucket = AwsArchitecture.getResource("AWS::S3::Bucket");
+
+const s3BucketPng = s3Bucket.icon("png"); // => "storage/s3/bucket.png"
````
should the image types be typed i.e. enumified?

JeremyJonas replied:
They are typed as a string literal union. The reason being is the jsii enums don't allow literal value comparison, and wanted to not require importing an enum for this.

https://aws.github.io/jsii/specification/2-type-system/#enum

JeremyJonas replied:
Discussed offline - resolving

agdimech commented on packages/aws-arch/scripts/common.ts:
```diff
@@ -0,0 +1,29 @@
+// Ensure the generated dir exists
+fs.ensureDirSync(GENERATED_DIR);
+
+export function logCount(title: string, total: number | object | any[] | Set<any>, count: number | object | any[] | Set<any>, delta: boolean = false): void {
```
can we make the type for total a type variable and re-use it here and on line 18?

agdimech commented on packages/aws-arch/scripts/common.ts:
```diff
@@ -0,0 +1,29 @@
+}
+
+export function parseCount(value: number | object | any[] | Set<any>): number {
+	if (value instanceof Set) {
```
convert these to if/else as I am not sure if a Set is also represented as an object and hence it may be possible to run multiple if conditions given we are not returning within the closure.

agdimech commented on packages/aws-arch/scripts/generate-assets.ts:
```diff
@@ -0,0 +1,392 @@
+#!/usr/bin/env ts-node
```
missing ASL header.

agdimech commented on packages/aws-arch/scripts/common.ts:
```diff
@@ -0,0 +1,29 @@
+import * as fs from 'fs-extra';
```
Missing ASL header. Not sure why the header eslint plugin didn't add this automatically.

agdimech commented on packages/aws-arch/scripts/generate-assets.ts:
```diff
@@ -0,0 +1,392 @@
+  template: 'management_governance/cloudformation/template',
+ } as const;
+
+export async function generate () {
```
Please add jsdocs to all public functions.

agdimech replied:
Still missing docs.

agdimech commented on packages/aws-arch/scripts/generate-assets.ts:
```diff
@@ -0,0 +1,392 @@
+
+const IOT_THING_PATTERN = /internet_of_things\/iot_thing\/(?<thingType>\w+)$/i;
+
+export const ASSET_SIZE = 450;
```
Please add jsdocs to all public consts.

agdimech commented on packages/aws-arch/scripts/generate-assets.ts:
```diff
@@ -0,0 +1,392 @@
+  await fs.copy(STATIC_ASSETS, ASSET_PACKAGE_DIR, { overwrite: true, recursive: true });
+
+  const cachedir = path.join(process.cwd(), 'node_modules', '.cache', '@aws-prototyping-sdk', 'aws-arch');
+  await fs.ensureDir(cachedir);
```
Does this still work if you blow away the cache and run generate?

JeremyJonas replied:
Yes - I just verified locally. Caching is just for downloaded files and deleting the cache works with subsequent generate

agdimech commented on packages/aws-arch/scripts/generate-assets.ts:
```diff
@@ -0,0 +1,392 @@
+
+  const cachedir = path.join(process.cwd(), 'node_modules', '.cache', '@aws-prototyping-sdk', 'aws-arch');
+  await fs.ensureDir(cachedir);
+  const downloapPath = path.join(cachedir, ASSET_PACKAGE_ZIP_FILENAME);
```
variable name has typo.

agdimech commented on packages/aws-arch/scripts/generate-assets.ts:
```diff
@@ -0,0 +1,392 @@
+  template: 'management_governance/cloudformation/template',
+ } as const;
+
+export async function generate () {
```
Is there a way to split this function up into multiple functions as its pretty hefty atm?

JeremyJonas replied:
The local variables are reused throughout, so while splitting it is definitely possible not sure it would become more readable nor maintainable at this point. Before this package become "stable" will definitely look into splitting it out, but not sure would get value for effort at this state. Happy to discuss and align if still a concern.

agdimech commented on packages/aws-arch/scripts/generate-mappings.ts:
```diff
@@ -0,0 +1,216 @@
+#!/usr/bin/env ts-node
```
Missing ASL header.

agdimech commented on packages/aws-arch/scripts/generate-drawio-spec.ts:
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env ts-node
```
Missing ASL header.

agdimech commented on packages/aws-arch/scripts/generate-cfnspec-types.ts:
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env ts-node
```
Missing ASL header.

agdimech commented on packages/aws-arch/scripts/generate-pricing-manifest.ts:
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env ts-node
```
Missing ASL header.

agdimech commented on packages/aws-arch/src/contants.ts:
```diff
@@ -0,0 +1,28 @@
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ******************************************************************************************************************** */
+export const ASSET_DIRNAME = "assets";
```
All exported consts should have jsdocs

agdimech reviewed (commented):
Good start, just took a very quick first pass.

JeremyJonas committed:
perf(nx): prevent all rebuild when adding new package

When nx.json is updated it causes all packages to be rebuilt (as nx.json is part of hash).

The docs package implicit dependencies where stored in nx.json, rather than within
the docs package itself.

JeremyJonas committed:
fix(github/workflows): add `--frozen-lockfile` to build workflow yarn install

JeremyJonas commented:
@agdimech Pushed new version which should resolve all comments and commented with additional context where not addressed in code.

I have added jsdocs to all public entities, as well as most internals.

Ready for diligent review

agdimech commented on .github/workflows/build.yml:
```diff
@@ -22,10 +22,10 @@ jobs:
           repository: ${{ github.event.pull_request.head.repo.full_name }}
       - name: Install dependencies
-        run: yarn install --check-files
+        run: yarn install --check-files --frozen-lockfile
```
Good catch :)

agdimech commented on packages/aws-arch/ASSETS.md:
```diff
@@ -0,0 +1,2824 @@
+<!-- AUTO-GENERATED - DO NOT EDIT -->
+  # AWS Architecture Icons
```
Please use ## instead as using a single # causes issues when rendering the microsite.

agdimech commented on packages/aws-arch/ASSETS.md:
```diff
@@ -0,0 +1,2824 @@
+<!-- AUTO-GENERATED - DO NOT EDIT -->
+  # AWS Architecture Icons
+
+  This package generated normalized asset dictionary from [AWS Architecture Icons](https://aws.amazon.com/architecture/icons/).
```
generates a *

agdimech commented on packages/aws-arch/ASSETS.md:
```diff
@@ -0,0 +1,2824 @@
+
+  Asset paths follow `<category>/<service>/<resource>[.<theme>].<ext>` structure.
+
+  > eg: storage/s3/bucket.png, storage/s3/bucket.dark.png=
```
Remove trailing =

agdimech commented on packages/aws-arch/ASSETS.md:
```diff
@@ -0,0 +1,2824 @@
+
+---
+
+## Available assets
```
Is there much benefit to listing all the assets here?

JeremyJonas replied:
Since removing the assets directory from source code (now only part of lib artifact), this file gives users visibility into the available assets within this package. Otherwise would have to install the package and looking into node_modules, or clone the repo and generate assets to see what is available.

So I believe it is quite useful for end users.

agdimech commented on packages/aws-arch/README.md:
```diff
@@ -0,0 +1,90 @@
+# aws-arch (AWS Architecture)
```
same here, please use ## instead of #

agdimech commented on packages/aws-arch/scripts/common.ts:
```diff
@@ -0,0 +1,47 @@
+export type TCountable = number | object | any[] | Set<any>;
+
+export function logCount(title: string, total: TCountable, count: TCountable, delta: boolean = false): void {
+	total = parseCount(total);
```
For this method we are mutating total and count; is that intentional?

I generally prefer to use immutable data structures in order to prevent unintended side effects. Can this be converted to using immutable types or do defensive copying?

JeremyJonas replied:
`total` and `count` are local function parameters, mutating them does not mutate the source. Can create local variables to store the values we want to use in the function, but it would not change anything except take more memory for additional variables. Agree that limiting side-effects outside of a function is preferred, but you can look at a functions parameters as local variable within the function, and in no way cause side-effects by changing what they reference.

agdimech commented on packages/aws-arch/scripts/common.ts:
```diff
@@ -0,0 +1,47 @@
+export const GENERATED_DIR = path.resolve(__dirname, '..', 'src', 'generated');
+fs.ensureDirSync(GENERATED_DIR);
+
+export const CACHE_DIR = path.join(process.cwd(), 'node_modules', '.cache', '@aws-prototyping-sdk', 'aws-arch');
```
Missing jsdocs.

agdimech commented on packages/aws-arch/scripts/common.ts:
```diff
@@ -0,0 +1,47 @@
+export const CACHE_DIR = path.join(process.cwd(), 'node_modules', '.cache', '@aws-prototyping-sdk', 'aws-arch');
+fs.ensureDirSync(CACHE_DIR);
+
+export type TCountable = number | object | any[] | Set<any>;
```
Missing jsdocs.

agdimech commented on packages/aws-arch/scripts/common.ts:
```diff
@@ -0,0 +1,47 @@
+
+export type TCountable = number | object | any[] | Set<any>;
+
+export function logCount(title: string, total: TCountable, count: TCountable, delta: boolean = false): void {
```
Missing jsdocs.

agdimech commented on packages/aws-arch/scripts/common.ts:
```diff
@@ -0,0 +1,47 @@
+	console.info(`${title}: %i% (%i of %i)`, (count / total) * 100, count, total)
+}
+
+export function parseCount(value: TCountable): number {
```
Is this used anywhere else? If so, please add jsdocs otherwise remove export.

agdimech commented on packages/aws-arch/scripts/generate-assets.ts:
```diff
@@ -0,0 +1,405 @@
+import fetch from 'node-fetch';
+import { Parse as ZipParse } from 'unzipper';
+// @ts-ignore
+import sharp = require('sharp'); // eslint-disable-line @typescript-eslint/no-require-imports
```
can we use @types/sharp?

agdimech commented on packages/aws-arch/scripts/generate-assets.ts:
```diff
@@ -0,0 +1,405 @@
+import { NORMALIZED_SERVICE_VARIANTS } from "../src/internal/services/variants";
+import { ASSET_DIRNAME, CATEGORY_ICON, PNG_ASSET_SIZE, SERVICE_ICON } from "../src/contants";
+
+const ASSET_PACKAGE_ZIP_URL = 'https://d1.awsstatic.com/webteam/architecture-icons/q3-2022/Asset-Package_07312022.e9f969935ef6aa73b775f3a4cd8c67af2a4cf51e.zip';
```
TILI: Add docs here stating this const is what should be updated when bumping icon versions. May also be beneficial to add a developer hint to the README pointing to this so it's easy for them to update in future.

JeremyJonas replied:
Done in upcoming revision

agdimech commented on packages/aws-arch/scripts/generate-assets.ts:
```diff
@@ -0,0 +1,405 @@
+  assetFiles.sort();
+
+  /** eslint-disable */
+  await fs.writeFile(AWS_ASSETS_TS, `// AUTO-GENERATED - DO NOT EDIT
```
Do you think we should use the ts.generateSourceFile instead so we get type safety?

JeremyJonas replied:
Yes that would be great, will consider refactoring in next revision.

agdimech commented on packages/aws-arch/scripts/generate-assets.ts:
```diff
@@ -0,0 +1,405 @@
+  /** eslint-enable */
+};
+
+export function parseAssetKey (assetKey:string, serviceNames: Set<string>): { serviceName: string, resourceName?: string } {
```
missing jsdocs

JeremyJonas replied:
Not actually used, will remove

agdimech commented on packages/aws-arch/scripts/generate-assets.ts:
```diff
@@ -0,0 +1,405 @@
+
+function resolveServiceName(value: string): string {
+  value = normalizeIdentifier(value);
+  if (['kinesis'].includes(value)) {
```
why is this needed?

JeremyJonas replied:
Added following comment to code to clarify
> // Kinesis is referenced as service and family, generally kinesis as "service" maps to "kinesis data stream" however for asset
> // resolution of value we want "kinesis" as family not direct service, so need to ignore service variant mapping for asset case.

agdimech commented on packages/aws-arch/scripts/generate-assets.ts:
```diff
@@ -0,0 +1,405 @@
+}
+
+function patchZipEntryPath (entryPath: string): string {
+  // there are some non _ delimeted asset names
```
delimited*

JeremyJonas replied:
Missed this comment, will resolve for now since just comment

agdimech commented on packages/aws-arch/scripts/generate-cfnspec-types.ts:
```diff
@@ -0,0 +1,67 @@
+
+const CFNSPEC_TS = path.join(GENERATED_DIR, 'cfnspec.ts');
+
+export async function generate () {
```
missing jsdocs

agdimech commented on packages/aws-arch/scripts/generate-drawio-spec.ts:
```diff
@@ -0,0 +1,67 @@
+import fetch from 'node-fetch';
+import { CACHE_DIR, GENERATED_DIR } from './common';
+// @ts-ignore
+import xmlFlow = require('xml-flow'); // eslint-disable-line @typescript-eslint/no-require-imports
```
Can we use @types/xml-flow?

JeremyJonas replied:
`@types/xml-flow` is being used. `ts-node` was throwing an error before for this, but I just removed the `@ ts-ingore` and it worked without error. Will commit with ignore removed.

agdimech commented on packages/aws-arch/scripts/generate-drawio-spec.ts:
```diff
@@ -0,0 +1,67 @@
+
+const DRAWIO_SPEC_TS = path.join(GENERATED_DIR, 'drawio-spec.ts');
+
+export async function generate () {
```
missing jsdocs

agdimech commented on packages/aws-arch/scripts/generate-drawio-spec.ts:
```diff
@@ -0,0 +1,67 @@
+
+  flowStream.on("end", async () => {
+		/* eslint-disable */
+    await fs.writeFile(DRAWIO_SPEC_TS, `// AUTO-GENERATED - DO NOT EDIT
```
same here, should we use ts generateSource and use the AST builder?

JeremyJonas replied:
I think that is a good improvement for these generated files, but would like to consider for later revision if ok.

agdimech commented on packages/aws-arch/scripts/generate-mappings.ts:
```diff
@@ -0,0 +1,231 @@
+
+const MAPPINGS_TS = path.join(GENERATED_DIR, 'mappings.ts')
+
+export async function generate () {
```
missing jsdocs

agdimech commented on packages/aws-arch/src/internal/mapping/helpers.ts:
```diff
@@ -0,0 +1,160 @@
+  if (resourceName == null) {
+    try {
+      resourceName = resolveResourceName(cfn.resourceName, serviceName);
+    } catch {}
```
Do we need to catch and swallow the error? Why not just omit and have it bubble up?

JeremyJonas replied:
I will add the follow comment to the code explaining reason to just ignore

> // There are lots of low-level cfn resource definitions without mappings to other systems,
> // for this reason we just ignore unresolved resources without spamming the console or
>  // bubbling the error as this is expected in large percent of cases.

agdimech commented on packages/aws-arch/src/internal/resources/helpers.ts:
```diff
@@ -0,0 +1,66 @@
+    }
+  }
+
+  throw new Error(
```
I notice we swallow these errors in consuming code. Does it still make sense to throw or should we log instead?

JeremyJonas replied:
How to handle resource name resolution is up to the consumer, as commented above. Depending on the mapping it is often very common for this to throw an error and therefore there spamming the console does not add value.

agdimech commented on private/projects/aws-arch.ts:
```diff
@@ -0,0 +1,98 @@
+        "xml-flow",
+      ],
+      peerDeps: ["projen"],
+      bundledDeps: ["lodash", "traverse", "fs-extra"],
```
Do these need to be bundled? Are we exposing lodash, traverse or fs-extra in any interfaces?

JeremyJonas replied:
Good catch! No they are not, was left over from refactor of the other projects inflight. `traverse` is not even used in this package and the others are dev deps only.

agdimech commented on packages/aws-arch/README.md:
```diff
@@ -0,0 +1,90 @@
+
+The primary aim of this package is to provide a consistent mechanism for other libraries to retrieve naming and assets associated with CloudFormation resources, and by extension CDK resources.
+
+## Get Started
```
Do we have the build/generate process documented anywhere i.e: run `npx projen generate`?

JeremyJonas replied:
I will add this to the README.md under a "Development" heading

agdimech reviewed (dismissed):
Looks good!

agdimech commented:
Getting failure during the build due to mutations in the README :(
